README.md
# linux-tablet-area-tool

> A tool for visualising and modifying the tablet area of a Wacom device using setxwacom as a backend

## Features
 - dynamic device detection
 - mutiple device support
 - Osu! sensitivity to area conversion
 
## Screenshots

![screenshot1](https://github.com/stefaneusunt/linux-wacom-tablet-area-tool/blob/master/ss1.png)
![screenshot2](https://github.com/stefaneusunt/linux-wacom-tablet-area-tool/blob/master/ss2.png)

 
## Building the project

``` bash
# install dependencies
npm install

# build and run the project
./build-run.sh

# generate a package for the current platform
./package-app.sh
```

CHANGELOG.md
Changelog
=========

- 1.0.0 - Initial release
  - Features:
    - automatic device detection
    - abilty to modify area both visually and using actual device values
  - Known issues:
    - the device area size is not dynamically detected
    - sometimes the computation of the current setting fails
    - Osu sensitivity is present in UI, but no functionality is implemented
- 1.1.1 - Async update
  - Changes:
    - the underlying Wacom library was rewritten to be completely aynchronous,
removing any hiccups in performance and improving the device detection rate
    - added the ability to switch between multiple devices

electron-main.js
const {app, BrowserWindow} = require('electron');
const path = require('path');
const url = require('url');


// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
let win;

function createWindow () {
  // Create the browser window. //619
  win = new BrowserWindow({width: 619, height: 581});
  win.setResizable(false);

  // and load the index.html of the app.
  win.loadURL(url.format({
    pathname: path.join(__dirname, '/index.html'),
    protocol: 'file:',

    slashes: true
  }));

  // Open the DevTools.
  //win.webContents.openDevTools();
  win.setMenu(null);


  // Emitted when the window is closed.
  win.on('closed', () => {
    // Dereference the window object, usually you would store windows
    // in an array if your app supports multi windows, this is the time
    // when you should delete the corresponding element.
    win = null;
    app.quit();
  })
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);

src/lib/wacom_API.js
////////////////////////////////
//       OBSOLETE CODE        //
// Am rescris libraria sa fie //
//   asincrona in wacom.js    //
////////////////////////////////

// A small lib that will interact with the Wacom devices using the xsetwacom cli tool
const cp = window.require('child_process');
const NodeConsole = window.require('console');

// Prepare an actual console
const console = new NodeConsole.Console(window.process.stdout, window.process.stderr);

export const rconsole = console;

export function cmd_out (cmd) {
  // Executes cmd, logs the call and returns the output as a string
  console.log('[cmd log] ' + cmd);
  return cp.execSync(cmd).toString();
}

export function list_devices_names() {
  // Return a list of Wacom devices
  const out = cmd_out('xsetwacom --list devices');
  const re = /(.*?) *?\t/;
  const devs = [];
  for (let l of out.split('\n')) {
    const m = re.exec(l);
    // console.log(m);
    if (m) {
      devs.push(m[1].trim());
    }
  }
  return devs;
}

export function list_device_params (devname) {
  // Return a object with parameters and their current value for the give device
  const re = /Option *?"(.*?)" *?"(.*?)"/;
  const out = cmd_out(`xsetwacom --get "${devname}" all`);
  let opts = {};
  for (let l of out.split('\n')) {
    const m = re.exec(l); // math re against l and return a match list
    // console.log(m);
    if (m) {
      opts[m[1]] = m[2];
    }
  }
  return opts;
}

export function set_device_param(devname, prop, val) {
  // Set a parameter for a device, returning true if succeds, false otherwise
  const out = cmd_out(`xsetwacom --set "${devname}" "${prop}" "${val}"`)

}

const MaxAreas = {
  'Wacom Intuos PT S 2 Pen stylus': {width: 15200, height: 9500}
};

export class WacomDevice {
  constructor(devname) {
    this.DeviceName = devname;
    const params = list_device_params(devname);
    this._params_cache = params;
    if (MaxAreas[devname]) {
      // This device has a known area size, expose it
      this.AreaSize = MaxAreas[devname];
    }
    for (let p in params) {
      Object.defineProperty(this, p, {
        get: () => {
          return this._params_cache[p];
        },
        set: (val) => {
          set_device_param(this.DeviceName, p, val);
          this._params_cache[p] = val;
        },
        enumerable: true
      });
    }
  }
}

export function list_devices() {
  let devs = [];
  for (let d of list_devices_names()) {
    devs.push(new WacomDevice(d));
  }
  return devs;
}



src/lib/wacom.js
// Like the old Wacom API, but implemented properly and async

const {execFile} = window.require('child_process');
const NodeConsole = window.require('console');

// Prepare an actual console
const console = new NodeConsole.Console(window.process.stdout, window.process.stderr);

function cmd_out(args) {
  return new Promise((resolve, reject) => {
    // Stringify everything
    args = args.map(x => {return x.toString()});
    execFile(args[0], args.slice(1), (err, stdout, stderr) => {
      console.log(`[cmd] ${args.join(' ')}`);
      if (err) {
        console.log(`Error: ${err.toString()}`);
        reject(err);
      } else {
        resolve(stdout + stderr);
      }
    });
  });
}
/*
cmd_out(['ls', '-lh']).then((out) => {
  console.log(out.toUpperCase());
});*/
function list_devs() {
  return new Promise((resolve, reject) => {
    cmd_out(['xsetwacom', '--list']).then((out) => {
      const re = /(.*?) *?id: *?(\d+)/;
      let devs = [];
      for (let l of out.split('\n')) {
        const m = re.exec(l);
        if (m) {
          devs.push({name: m[1].trim(), id: m[2]});
        }
      }
      resolve(devs);
    });
  });
}


list_devs().then((devs) => {
  console.log('Found the following devices:\n', devs);
});

class WacomDevice {
  constructor(devname, id) {
    this.DeviceName = devname;
    this.id = id;
  }
  get_param(param_name) {
    return new Promise((resolve, reject) => {
      cmd_out(['xsetwacom', '--get', this.id, param_name]).then((val) => {
        val = val.trim();
        console.log(`[WacomDevice::${this.DeviceName}::${param_name}] -> '${val}'`);
        resolve(val);
      });
    });
  }
  set_param(param_name, val) {
    return new Promise((resolve, reject) => {
      cmd_out(['xsetwacom', '--set', this.id, param_name, val]).then((out) => {
        console.log(`[WacomDevice::${this.DeviceName}::${param_name}] <- '${val}'.`);
        resolve('ok');
      });
    });
  }
}

export function get_wacom_devices() {
  return new Promise((resolve, reject) => {
    list_devs().then((devs) => {
      Promise.all(devs.map(d => {
        return new Promise((resolve, reject) => {
          let wd = new WacomDevice(d.name, d.id);
          // Now get the maximum area if possible
          cmd_out(['xinput', '--list', d.id]).then((xinput_out) => {
            const re = /Label: (.+?)\n\s*?Range: (.+)\n/g;
            let opts = {};
            let m;
            do {
              m = re.exec(xinput_out);
              if (m) {
                opts[m[1]] = m[2];
              }
            } while (m);
            // all xinput parameters have been parsed
            // compute the max area size and add it to the wacom device obj
            // console.log(`${d.name}: ${JSON.stringify(opts)}`);
            if (opts['Abs X'] && opts['Abs Y']) {
              // gott'em
              const [w, h] = [opts['Abs X'], opts['Abs Y']].map(x => {
                return parseInt(x.split('-')[1].trim().split('.')[0]);
              });
              wd.AreaSize = {width: w, height: h}
            }
            resolve(wd);
          });
        });
      })).then((devs) => {
        devs.sort();
        resolve(devs)});
    });
  });
}


get_wacom_devices().then(devs => {console.log(devs)});
/*
get_wacom_devices.then((devs) => {
  for (d of devs) {
    if (d.DeviceName.endsWith('stylus')) {
      d.get_param('Area').then((area) => {
        console.log(`Tablet area: ${area}`);
        return d.set_param('Area', '3000 3000 15200 9500');
      }).then(() => {
        console.log('set');
      })
    }
  }
});*/